File: 97 (Interleaving String)/python/Solution.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:

        if (len(s1)+len(s2)) != len(s3): return False

        @cache
        def solve(i,j):
            if i==len(s1) and j == len(s2): return True

            a,b= False,False
            if i<len(s1) and s1[i]==s3[i+j]:
                a = solve(i+1,j)

            if j<len(s2) and s2[j]==s3[i+j]:
                b = solve(i,j+1)

            return a or b
        
        return solve(0,0)

File: 97 (Interleaving String)/java/Solution.java
class Solution {

    public boolean solve(int i, int j, String s1, String s2, String s3, int[][] dp){
        if (i==s1.length() && j==s2.length()) return true;

        if (dp[i][j]!=0){
            return dp[i][j]==1;
        }

        boolean a = false, b=  false;
        if (i<s1.length() && s1.charAt(i)==s3.charAt(i+j)){
            a = solve(i+1,j,s1,s2,s3,dp);
        }
        if (j<s2.length() && s2.charAt(j)==s3.charAt(i+j)){
            b = solve(i,j+1,s1,s2,s3,dp);
        }
        dp[i][j] = (a||b) ? 1 : -1;
        return dp[i][j]==1;
        
    }
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;

        int n = s1.length();
        int m = s2.length();
        int dp[][]  = new int[n+1][m+1];

        return solve(0,0, s1,s2,s3,dp);
    }
}